web/handlers/auth/withssb.go

```diff
@@ -0,0 +1,115 @@
+// SPDX-License-Identifier: MIT
+
+package auth
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"net/http"
+	"time"
+
+	"go.cryptoscope.co/muxrpc/v2"
+
+	"go.mindeco.de/http/auth"
+	refs "go.mindeco.de/ssb-refs"
+
+	"github.com/ssb-ngi-pointer/go-ssb-room/internal/network"
+	"github.com/ssb-ngi-pointer/go-ssb-room/internal/signinwithssb"
+	"github.com/ssb-ngi-pointer/go-ssb-room/roomdb"
+	weberrors "github.com/ssb-ngi-pointer/go-ssb-room/web/errors"
+)
+
+// withssbHandler implements the oauth-like challenge/response dance described in
+// https://ssb-ngi-pointer.github.io/rooms2/#sign-in-with-ssb
+type withssbHandler struct {
+	roomID refs.FeedRef
+
+	members roomdb.MembersService
+	aliases roomdb.AliasesService
+
+	cookieAuth *auth.Handler
+
+	endpoints network.Endpoints
+}
+
+func (h withssbHandler) login(w http.ResponseWriter, req *http.Request) (interface{}, error) {
+	queryParams := req.URL.Query()
+
+	var clientReq signinwithssb.ClientRequest
+	clientReq.ServerID = h.roomID // fll inthe server
+
+	// validate and update client challange
+	cc := queryParams.Get("challenge")
+	if _, err := signinwithssb.DecodeChallengeString(cc); err != nil {
+		return nil, weberrors.ErrBadRequest{Where: "client-challange", Details: err}
+	}
+	clientReq.ClientChallange = cc
+
+	// check who the client is
+	var client refs.FeedRef
+	if cid := queryParams.Get("cid"); cid != "" {
+		parsed, err := refs.ParseFeedRef(cid)
+		if err != nil {
+			return nil, weberrors.ErrBadRequest{Where: "cid", Details: err}
+		}
+		client = *parsed
+	} else {
+		alias, err := h.aliases.Resolve(req.Context(), queryParams.Get("alias"))
+		if err != nil {
+			return nil, weberrors.ErrBadRequest{Where: "alias", Details: err}
+		}
+		client = alias.Feed
+	}
+
+	// check that we have that member
+	member, err := h.members.GetByFeed(req.Context(), client)
+	if err != nil {
+		if err == roomdb.ErrNotFound {
+			return nil, weberrors.ErrForbidden{Details: fmt.Errorf("sign-in: client isnt a member")}
+		}
+		return nil, err
+	}
+	clientReq.ClientID = client
+
+	// get the connected client for that member
+	edp, connected := h.endpoints.GetEndpointFor(client)
+	if !connected {
+		return nil, weberrors.ErrForbidden{Details: fmt.Errorf("sign-in: client not connected to room")}
+	}
+
+	// roll a challange from the server
+	sc := signinwithssb.GenerateChallenge()
+	clientReq.ServerChallange = sc
+
+	ctx, cancel := context.WithTimeout(req.Context(), 1*time.Minute)
+	defer cancel()
+
+	// request the signed solution over muxrpc
+	var solution string
+	err = edp.Async(ctx, &solution, muxrpc.TypeString, muxrpc.Method{"httpAuth", "requestSolution"}, sc, cc)
```

Comment:
Perfect. This matches https://github.com/ssb-ngi-pointer/ssb-http-auth-client/blob/25430acb55c5466fed4cd245b0997c6f4d889227/src/plugin-http-auth.ts#L27